### PR TITLE
[WIP] Auto-document JSON API with Swagger

### PR DESCRIPTION
**Pull request by hoffmabc**
I'm going to slowly hammer away at this migration to self-documenting Swagger docs for our API in the hopes that we can work to keep our API well documented and up to date. The Swagger doc can be generated automatically during the Travis builds when we publish and then can be imported into Postman if we like. I've begun porting over the current Postman docs to the golang code. It will take a bit of time to catch it up, but once it's in there it should be straightforward to maintain.

Thoughts @cpacia @placer14 @tyler-smith ?

### CODE REVIEW

**Comment by hoffmabc** [comment]
If you want to see what it generates locally you can do the following to test it out:

```
cd api/
swaggegenerate spec -b . -o ./swagger.json --scan-models
swagger flatten swagger.json > flatten.json
swagger serve -F=swagger flatten.json
```

image

File: api/jsonapi.go [hunk 1]
```diff
@@ -1,3 +1,27 @@
...
...s <em>base64encode(username + \":\" + password)</em><br>or included in the Url:<br><code><br>http://username:password@localhost\:8080/ob/<br></code><br># SSL<br><strong>NEVER</strong> open the JSON API to the internet without also encrypting the connection with SSL. Without SSL enabled your authentication information be sent to the server in the clear allowing anyone who views your packets to access your server and potentially steal your bitcoins. For instructions on how to enable SSL, see [here](https://github.com/OpenBazaar/openbazaar-go/blob/master/docs/ssl.md).<br># Cross-Origin Resource Sharing<br>CORS is turned off by default, meaning that you will not be able to make API calls from a browser to a running server on localhost. You can turn CORS on for a specific domain by setting an origin in the config file as follows:<br><code><br>{<br>    \"JSON-API\": {<br>      \"CORS\": \"https://openbazaar.org\"<br>      }<br>}<br></code><br>Use \"*\" to allow all domains. Keep in mind that running the server with CORS enabled for all domains is a potential secrity risk as it will allow any website you visit to make calls to your server. Authentication should be enabled along side CORS.
```
Comment (placer14) [inline api/jsonapi.go]: Gross.

Comment (placer14) [reply api/jsonapi.go]: Could this be formatted more nicely?

Comment (hoffmabc) [reply api/jsonapi.go]: It could be, but I'm directly porting from Postman. Thoughts @drwasho ?

**Comment by christroutner** [comment]
Before you guys commit to one path forward with regard to documentation, allow me to point out some alternatives.

* [mdoc](https://github.com/millermedeiros/mdoc) is a markdown generator, that I use to write [the protocol specifications for P2P VPS](https://p2pvps.net/doc/index.html). Markdown, updated via GitHub PRs, is a great way to manage long-form documentation. The PR system lets you leverage your community to make updates and keep the documentation up to date.

* [apidocs](http://apidocjs.com/) is an auto-generated documentation I use to document [the P2P VPS server API](https://p2pvps.net/docs/). This is what I think you're trying to achieve with the OB API. The documentation lives side-by-side with the code, so it's easy to update the documentation as the code gets updated.

These options are obviously targeted to a JavaScript environment, and you guys are using Go, but I'm sure there are Go-focused alternatives that function in a similar manner.

^^^ @hoffmabc @drwasho @SamPatt @placer14 @cpacia @tyler-smith

**Comment by placer14** [comment]
@christroutner We're exploring still and appreciate the links. But let's keep conversation in this PR specific to this swagger example. I've made an issue (https://github.com/OpenBazaar/openbazaar-go/issues/999) where we can hash over the pros/cons if we like. Would you like to edit your comment to add any pros/cons that we should consider and move it into the issue? Further, examples that we can size up along side swagger are even more valuable. I know @BazaarDog has also done some work here. I invite you both to share live examples in the issue thread for dissection.

**Comment by placer14** [comment]
We decided to continue improving the Postman documentation and incorporate them into travis CI.

**Pull request closed without merging**